### PR TITLE
AP_HAL_ChibiOS: add two flight controllers ACNS_CM4PILOT and ACNS_F405AIO

### PR DESCRIPTION
**Pull request by robinluojl**
AP_HAL_ChibiOS: add two flight controllers ACNS_CM4PILOT and ACNS_F405AIO

### CODE REVIEW

**Comment by robinluojl** [comment]
we will update README.md as soon as possible. pictures of these products will update later.

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+| PWM4         | 53   | 
+| PWM5         | 54   | 
+| PWM6         | 55   | 
+| PWM7         | 56   | 
+
+## Analog inputs
+
+The CM4PILOT flight controller has 3 analog inputs
+
+ - ADC Pin11  -> Battery Current
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/README.md:61]: Need scales....see https://github.com/ArduPilot/ardupilot/tree/master/libraries/AP_HAL_ChibiOS/hwdef/SpeedyBeeF405WING for an example of what is needed in the README.md

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ - 2 UARTs and 4 USB2.0 and 1 OTG for CM4
+ - 1 I2C port
+ - 1 CAN port
+ - 1 SBUS input and 8 PWM output (all support DShot)
+ - External SPI port
+ - Buzzer on board 
+ - GRB LED on board
+ - 128M flash on board
+
+## UART Mapping
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/README.md:33]: need DMA capability for each port

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+ - SERIAL0 -> USB(OTG1)
+ - SERIAL1 -> USART1(Telem1)
+ - SERIAL2 -> USART3(Telem2)
+ - SERIAL3 -> UART4(GPS)
+ - SERIAL4 -> UART6(GPS2)
+ - SERIAL5 -> USART2(SBUS)
+ 
+## GPIOs
+
+4 External PWM channels can be used for GPIO functions.
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/README.md:33]: need grouping and capability (BDShot? or just DShot>)

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/README.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+| PWM Channels | Pin  |
+| ------------ | ---- | 
+| PWM5         | 54   |
+| PWM6         | 55   | 
+| PWM7         | 56   | 
+| PWM8         | 57   | 
+
+## Analog inputs
+
+The ACNS_F405AIO flight controller has 3 analog inputs
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/README.md:45]: Need Monitor/Scale info

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/README.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+ - ADC Pin11  -> Battery Current 
+ - ADC Pin12  -> Battery Voltage 
+ - ADC Pin10  -> RSSI voltage monitoring
+
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/README.md:50]: Need other sections as per example link given above

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/README.md (hunk 3 above)
Comment (robinluojl) [reply libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/README.md:33]: just Dshot

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/Readme.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
+# CM4PILOT Integrated Flight Controller
```
[the remark this answers is not in the document]
Comment (robinluojl) [reply libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/Readme.md:1]: yes it is. how to delete it

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/F405AIO.ioc [hunk 7]
```diff
@@ -0,0 +1,321 @@
+#MicroXplorer Configuration settings - do not modify
```
[the remark this answers is not in the document]
Comment (robinluojl) [reply libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/F405AIO.ioc:1]: this is a stm32cubeMX configuraiton file.  it's unnecessary for wiki.  F405AIO can be working with any kind of companion computer via one UART port. we are working on CM4Pilot configuration and our website, but we need more time.

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/README.md (hunk 1 above)
Comment (robinluojl) [reply libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/README.md:61]: I will learn and add the missing information

Comment (robinluojl) [reply libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/README.md:61]: thanks a lot

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/Readme.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/ACNS-CM4pilot/Readme.md:1]: go to your branch, delete the file, commit change, push back to your origin like you did to submit the PR

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/F405AIO.ioc (hunk 7 above)
[the remark this answers is not in the document]
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/F405AIO.ioc:1]: its also unnecessary in the HWDEF folder then....you should remove it

[the remark this answers is not in the document]
Comment (robinluojl) [reply libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/F405AIO.ioc:1]: ok， it will be removed

File: Tools/AP_Bootloader/board_types.txt [hunk 8]
```diff
@@ -225,6 +225,8 @@ AP_HW_SDMODELH7V1                    1111
 AP_HW_FlyingMoonH743                 1112
 AP_HW_YJUAV_A6                       1113
 AP_HW_YJUAV_A6Nano                   1114
+AP_HW_ACNS_CM4PILOT                  1115
```
Comment (peterbarker) [inline Tools/AP_Bootloader/board_types.txt:228]: If you want to reserve these IDs, make a separate PR.  Otherwise someone might get in ahead of you.

File: libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/hwdef.dat [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+define STM32_PWM_USE_ADVANCED TRUE
+
+
+
+# --------------------- save flash ----------------------
+include ../include/minimize_features.inc
+# only include ublox GPS driver
+#include ../include/minimal_GPS.inc
+define AP_PARAM_MAX_EMBEDDED_PARAM 1024
+#define HAL_GENERATOR_ENABLED 0
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/ACNS-F405AIO/hwdef.dat:208]: Please don't enumerate things in here, apart from parachute.  Use the save_some_flash.inc and `minimal.inc` instead.

**Review by peterbarker — CHANGES_REQUESTED** [review]
This should only be a single commit - please squash all your commits together.  Instructions here: https://ardupilot.org/dev/docs/git-interactive-rebase.html

Force-push the result over this branch.

**Comment by Hwurzburg** [comment]
@robinluojl are you ready for me to re-review? you also need to add the bootloaders ,squash to one commit and then library split the commits to Tools and hwdef..of course if you haven't reserved the board ids, you may have to redo those lines and the bootloaders....suggest you do a PR now for the ids...

**Comment by robinluojl** [comment]
@Hwurzburg I will do those things one by one。 I have started a new  PR now for the IDs.  Thanks

**Comment by Hwurzburg** [comment]
@robinluojl if you need help on squash and library split, I can share screen with you and walk you thru it on Discord if you wish...I am in Austin,TX time zone
